docs/tutorial/2_MarketMaker.md

```diff
@@ -0,0 +1,230 @@
+
+
+```
+S1 = 1540701333592592592592614116
+S2 = 1e24
+ ```
+
+I inserted 1000 CLN to the R1 reservoir and got back ~1139 CC from R2 reserve. That's say:
+
+```
+R1 = 1e21
+R2 = 1e24 - 1,139.346173578872162244e18 = 9.988606538264211e23
+```
+
+
+After the substitution we see that the first part of the formula is really close to 1 and the second part close to 0:
+
+```
+~1 + ~0 = 1
+```
+Q.E.D. :white_check_mark:
+
+
+### Current Price Formula
+
+Showing you just the derived formula for `getCurrentPrice`, it is:
+
+![price](../assets/formulas/price.gif)
+
+Oh don't ask me how we did this (read the whitepaper). The important thing here is because S1 and S2 are constants, price is expressed as a function of R1. And this is great because, as you know, one variable functions are simple and easy to use.
+
+##### getCurrentPrice_before
+
+That's the first time we checked `getCurrentPrice`. I inserted 1000 CLN before so all variables from last section still hold. I substitute the variables and calculate:
+
+```
+price = 0.5696729019147757
+```
+
+Comparing this price to a `getCurrentPrice` answer, we're getting a really close numbers. I did my calculations using python, which uses floating point to store real numbers, while Solidity stores all numbers as natural ones. This makes Solidity's calculation more accurate than mine. Python even doesn't see the difference between the two results, but calculating the diff in [wolframalpha](http://www.wolframalpha.com/input/?i=569672901914775677+%2F+1e18+-+0.5696729019147757) I can see that the diff is `-2.3e-17`. I hope it shows that the contract's calculations are more accurate that mine :sweat_smile:.
+
+ I think you're getting the point and can calculate `getCurrentPrice_after` by yourself :wink:. But I want to assure you that the price is much more stable than we saw. `getCurrentPrice` returns the exchange rate of CLN to CC, as we saw that gave us  `1 CLN => ~0.57 CC`. I wish to present a reverse exchange of CC to CLN, to show CC's growth of value. The formula for that is just `1 / getCurrentPrice`, now let's see some graphs I've drawn.
+
+ CC Price with CLN reservoir from 1 to 10000 CLN:
+
+ ![CC_price](../assets/graphs/CC_price.png)
+
+As you see, initially the price climbs steeply, but then the amplitude is kind of stays the same.
+
+That's the same function over the domain of 1 to 1e6:
+
+![CC_price2](../assets/graphs/CC_price2.png)
+
+It's important to note that the maximum for CLN reserve is S1 and not 1e6, but you see the pattern. I call this pretty stable.
+
+### Calculating Quote
+
+As we saw before, `quote` function calculates how much of token1 received in exchange for certain amount of token2. For now we are interested in exchange of CLN to CC, but the CC to CLN exchange calculated in the same way. Let's follow `quote` logic and calculate ourselves how much CLN I received in return.
+
+If we poke a little the ellipse equation we can present a formula for R2:
```

Comment:
the ellipse equation --> in the ellipse equation

Reply:
It's like "poke the bear"